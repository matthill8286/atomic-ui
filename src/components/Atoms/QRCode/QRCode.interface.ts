import { saiyanTheme } from '@/styles/sc-vars-saiyan'

export type ThemeColors = keyof typeof saiyanTheme.color & string

export enum RenderAs {
  CANVAS = 'canvas',
  SVG = 'svg',
}

/** QR code error correction levels */
export enum ErrorLevel {
  /** L - 7% of codewords can be restored */
  LOW = 'L',
  /** M - 15% of codewords can be restored */
  MEDIUM = 'M',
  /** Q - 25% of codewords can be restored */
  QUARTILE = 'Q',
  /** H - 35% of codewords can be restored */
  HIGH = 'H',
}

export interface QRCodeProps {
  value: string
  renderAs?: RenderAs
  size?: number
  errorLevel?: ErrorLevel
  includeMargin?: boolean
}
