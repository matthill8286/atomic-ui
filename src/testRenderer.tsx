import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
