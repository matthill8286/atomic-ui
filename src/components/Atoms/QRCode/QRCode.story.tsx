import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { StorybookWrapper } from '@/utils/StorybookWrapper'
import { QRCodeGenerator } from './QRCode'
import { ErrorLevel, RenderAs } from './QRCode.interface'

storiesOf('Design System/Atoms/QRCode', module)
  .add('Default', () => {
    return (
      <StorybookWrapper>
        <QRCodeGenerator value="9802149999997839" />
      </StorybookWrapper>
    )
  })
  .add('Render as Canvas', () => {
    return (
      <StorybookWrapper>
        <QRCodeGenerator value="9802149999997839" renderAs={RenderAs.CANVAS} />
      </StorybookWrapper>
    )
  })
  .add('Customized Size', () => {
    return (
      <StorybookWrapper>
        <QRCodeGenerator value="9802149999997839" size={256} />
      </StorybookWrapper>
    )
  })
  .add('With Error', () => {
    return (
      <StorybookWrapper>
        <QRCodeGenerator value="9802149999997839" errorLevel={ErrorLevel.HIGH} />
      </StorybookWrapper>
    )
  })
  .add('With Margin', () => {
    return (
      <StorybookWrapper>
        <QRCodeGenerator value="9802149999997839" includeMargin />
      </StorybookWrapper>
    )
  })
