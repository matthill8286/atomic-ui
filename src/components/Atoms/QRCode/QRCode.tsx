import QRCode from 'qrcode.react'
import * as React from 'react'
import { ThemeColors } from '@/types/theme'
import { ErrorLevel, QRCodeProps, RenderAs } from './QRCode.interface'

const bgColor: ThemeColors = 'white'
const fgColor: ThemeColors = 'black'

export const QRCodeGenerator: React.FC<QRCodeProps> = ({
  value,
  renderAs = RenderAs.SVG,
  size = 128,
  errorLevel = ErrorLevel.LOW,
  includeMargin = false,
}) => {
  return (
    <QRCode
      value={value}
      renderAs={renderAs}
      size={size}
      bgColor={bgColor}
      fgColor={fgColor}
      level={errorLevel}
      includeMargin={includeMargin}
    />
  )
}
