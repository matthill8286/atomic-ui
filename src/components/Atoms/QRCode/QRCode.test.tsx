import React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { QRCodeGenerator } from './QRCode'
import { ErrorLevel, RenderAs } from './QRCode.interface'

describe('QRCode', () => {
  it('renders for the input with all attributes', () => {
    const tree = renderWithTheme(
      <QRCodeGenerator
        value="9802149999997839"
        renderAs={RenderAs.SVG}
        size={256}
        errorLevel={ErrorLevel.QUARTILE}
        includeMargin
      />
    )
    expect(tree).toMatchSnapshot()
  })
})
