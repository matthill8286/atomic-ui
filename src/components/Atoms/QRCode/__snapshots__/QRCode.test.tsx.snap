// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCode renders for the input with all attributes 1`] = `
<svg
  height={256}
  shapeRendering="crispEdges"
  viewBox="0 0 33 33"
  width={256}
>
  <path
    d="M0,0 h33v33H0z"
    fill="white"
  />
  <path
    d="M4 4h7v1H4zM13 4h3v1H13zM17 4h3v1H17zM22,4 h7v1H22zM4 5h1v1H4zM10 5h1v1H10zM16 5h1v1H16zM19 5h1v1H19zM22 5h1v1H22zM28,5 h1v1H28zM4 6h1v1H4zM6 6h3v1H6zM10 6h1v1H10zM14 6h1v1H14zM16 6h2v1H16zM22 6h1v1H22zM24 6h3v1H24zM28,6 h1v1H28zM4 7h1v1H4zM6 7h3v1H6zM10 7h1v1H10zM12 7h1v1H12zM15 7h2v1H15zM19 7h1v1H19zM22 7h1v1H22zM24 7h3v1H24zM28,7 h1v1H28zM4 8h1v1H4zM6 8h3v1H6zM10 8h1v1H10zM14 8h1v1H14zM19 8h2v1H19zM22 8h1v1H22zM24 8h3v1H24zM28,8 h1v1H28zM4 9h1v1H4zM10 9h1v1H10zM12 9h2v1H12zM16 9h3v1H16zM20 9h1v1H20zM22 9h1v1H22zM28,9 h1v1H28zM4 10h7v1H4zM12 10h1v1H12zM14 10h1v1H14zM16 10h1v1H16zM18 10h1v1H18zM20 10h1v1H20zM22,10 h7v1H22zM12 11h1v1H12zM17 11h1v1H17zM5 12h2v1H5zM10 12h1v1H10zM13 12h2v1H13zM16 12h2v1H16zM22 12h2v1H22zM25 12h1v1H25zM4 13h1v1H4zM6 13h4v1H6zM12 13h1v1H12zM14 13h1v1H14zM16 13h1v1H16zM19 13h1v1H19zM21 13h2v1H21zM27,13 h2v1H27zM6 14h2v1H6zM9 14h2v1H9zM12 14h1v1H12zM14 14h6v1H14zM21 14h1v1H21zM24,14 h5v1H24zM4 15h3v1H4zM11 15h1v1H11zM13 15h4v1H13zM18 15h2v1H18zM21 15h1v1H21zM24 15h1v1H24zM5 16h3v1H5zM10 16h2v1H10zM13 16h2v1H13zM17 16h3v1H17zM21 16h2v1H21zM27 16h1v1H27zM5 17h3v1H5zM14 17h1v1H14zM16 17h4v1H16zM21 17h2v1H21zM27,17 h2v1H27zM4 18h2v1H4zM8 18h3v1H8zM12 18h1v1H12zM14 18h1v1H14zM19 18h3v1H19zM24 18h2v1H24zM27,18 h2v1H27zM7 19h2v1H7zM11 19h3v1H11zM16 19h3v1H16zM21 19h1v1H21zM24 19h1v1H24zM27 19h1v1H27zM4 20h4v1H4zM9 20h2v1H9zM13 20h1v1H13zM15 20h3v1H15zM20 20h5v1H20zM27 20h1v1H27zM12 21h3v1H12zM16 21h1v1H16zM19 21h2v1H19zM24 21h1v1H24zM27,21 h2v1H27zM4 22h7v1H4zM14 22h1v1H14zM16 22h3v1H16zM20 22h1v1H20zM22 22h1v1H22zM24 22h1v1H24zM26,22 h3v1H26zM4 23h1v1H4zM10 23h1v1H10zM14 23h2v1H14zM17 23h1v1H17zM20 23h1v1H20zM24 23h1v1H24zM4 24h1v1H4zM6 24h3v1H6zM10 24h1v1H10zM13 24h1v1H13zM17 24h1v1H17zM19 24h6v1H19zM27,24 h2v1H27zM4 25h1v1H4zM6 25h3v1H6zM10 25h1v1H10zM13 25h2v1H13zM16 25h5v1H16zM23 25h3v1H23zM27 25h1v1H27zM4 26h1v1H4zM6 26h3v1H6zM10 26h1v1H10zM12 26h1v1H12zM14 26h2v1H14zM17 26h2v1H17zM20 26h1v1H20zM23,26 h6v1H23zM4 27h1v1H4zM10 27h1v1H10zM12 27h1v1H12zM15 27h2v1H15zM19 27h2v1H19zM22 27h2v1H22zM4 28h7v1H4zM13 28h1v1H13zM15 28h1v1H15zM17 28h1v1H17zM19 28h2v1H19zM22 28h2v1H22zM28,28 h1v1H28z"
    fill="black"
  />
</svg>
`;
